Add option to print word-level scores

snukky opened the pull request:
This adds two options to marian-decoder:
- `--word-scores` displays (not normalized) word-level scores calculated from the path score.
- `--no-spm-decode` keeps output BPE-segmented, available only if Marian is compiled with SentencePiece. 

The latter may be handy if used with `--word-scores`/`--alignment` to get 1:1 mapping between output tokens and scores/alignment.

Examples:
```
echo "This is a test." | marian-decoder -c config.yml -n 1 --word-scores
Tohle je test. ||| WordScores= -1.51923 -0.21951 -1.48668 -0.24813 -0.22176

echo "This is a test." | marian-decoder -c config.yml -n 1 --word-scores --n-best
0 ||| Tohle je test. ||| F0= -1.77547 F1= -1.91984 WordScores= -1.51923 -0.21951 -1.48668 -0.24813 -0.22176 ||| -0.739061
0 ||| Tohle je zkouška. ||| F0= -2.09314 F1= -2.03168 WordScores= -1.51923 -0.21951 -1.92797 -0.23666 -0.22146 ||| -0.824964
0 ||| To je test. ||| F0= -3.32969 F1= -3.23537 WordScores= -4.19787 -0.27764 -1.61160 -0.25529 -0.22266 ||| -1.31301

echo "This is a test." | marian-decoder -c config.yml -n 1 --word-scores --n-best --no-spm-decode
0 ||| ▁Tohle ▁je ▁test . </s> ||| F0= -1.77547 F1= -1.91984 WordScores= -1.51923 -0.21951 -1.48668 -0.24813 -0.22176 ||| -0.739061
0 ||| ▁Tohle ▁je ▁zkouška . </s> ||| F0= -2.09314 F1= -2.03168 WordScores= -1.51923 -0.21951 -1.92797 -0.23666 -0.22146 ||| -0.824964
0 ||| ▁To ▁je ▁test . </s> ||| F0= -3.32969 F1= -3.23537 WordScores= -4.19787 -0.27764 -1.61160 -0.25529 -0.22266 ||| -1.31301

echo "This is a test." | marian-decoder -c config.yml -n 1 --word-scores --no-spm-decode --alignment
▁Tohle ▁je ▁test . </s> ||| 1-0 5-1 5-2 5-3 5-4 ||| WordScores= -1.51923 -0.21951 -1.48668 -0.24813 -0.22176
```

emjotde commented on src/data/sentencepiece_vocab.cpp:
```diff
@@ -36,6 +36,8 @@ class SentencePieceVocab : public IVocab {
   std::mt19937 generator_;
   std::uniform_int_distribution<int> randInt_; // from 0 to INT_MAX
 
+  bool keepEncoded_{false};
```
Please add a comment what this does.

snukky replied:
Added

emjotde commented on src/data/sentencepiece_vocab.cpp:
```diff
@@ -223,11 +227,18 @@ class SentencePieceVocab : public IVocab {
-    for (auto&& word : sentence)
-      spmSentence.push_back(word.toWordIndex());
-    spm_->Decode(spmSentence, &line);
+    if(keepEncoded_) {
```
Comment please.

snukky replied:
Done

emjotde commented on src/translator/output_printer.cpp:
```diff
@@ -26,4 +28,12 @@ std::string OutputPrinter::getAlignment(const Ptr<Hypothesis>& hyp) {
   }
 }
 
+std::string OutputPrinter::getWordScores(const Ptr<Hypothesis>& hyp) {
```
Comment what these are.

snukky replied:
I added a more detailed comment to the function generating the scores

emjotde commented:
Generally approved. I am a bit unsure about the missing `|||` between feature functions and word scores. Word scores cannot be used for MERT or similar optimizations.

snukky commented:
Added comments and changed the placement of word-level scores in the n-best list to a separate field before feature function scores.
With word alignment, the order is as follows: id, sentence, word alignment, word scores, feature functions, total score
